File: src/main/java/com/hileco/cortex/instructions/calls/CALL.java
package com.hileco.cortex.instructions.calls;

import com.hileco.cortex.context.ProcessContext;
import com.hileco.cortex.context.ProgramContext;
import com.hileco.cortex.context.ProgramZone;
import com.hileco.cortex.context.layer.Pair;
import com.hileco.cortex.instructions.Instruction;
import com.hileco.cortex.instructions.ProgramException;
import com.hileco.cortex.instructions.StackParameter;
import lombok.EqualsAndHashCode;

import java.math.BigInteger;
import java.util.List;

import static com.hileco.cortex.context.ProgramZone.MEMORY;
import static com.hileco.cortex.context.ProgramZone.PROGRAM_CONTEXT;
import static com.hileco.cortex.context.ProgramZone.STACK;
import static com.hileco.cortex.instructions.ProgramException.Reason.CALL_RECIPIENT_MISSING;
import static com.hileco.cortex.instructions.ProgramException.Reason.STACK_TOO_FEW_ELEMENTS;

@EqualsAndHashCode
public class CALL implements Instruction {

    public static final StackParameter RECIPIENT_ADDRESS = new StackParameter("recipientAddress", 0);
    public static final StackParameter VALUE_TRANSFERRED = new StackParameter("valueTransferred", 1);
    public static final StackParameter IN_OFFSET = new StackParameter("inOffset", 2);
    public static final StackParameter IN_SIZE = new StackParameter("inSize", 3);
    public static final StackParameter OUT_OFFSET = new StackParameter("outOffset", 4);
    public static final StackParameter OUT_SIZE = new StackParameter("outSize", 5);

    @Override
    public void execute(ProcessContext process, ProgramContext program) throws ProgramException {
        var stack = program.getStack();
        if (stack.size() < 6) {
            throw new ProgramException(program, STACK_TOO_FEW_ELEMENTS);
        }
        var recipientAddress = new BigInteger(stack.pop());
        var valueTransferred = new BigInteger(stack.pop());
        var inOffset = new BigInteger(stack.pop());
        var inSize = new BigInteger(stack.pop());
        var outOffset = new BigInteger(stack.pop());
        var outSize = new BigInteger(stack.pop());

        program.setReturnDataOffset(outOffset);
        program.setReturnDataSize(outSize);
        var recipient = process.getAtlas().get(recipientAddress);
        if (recipient == null) {
            throw new ProgramException(program, CALL_RECIPIENT_MISSING);
        }
        var sourceAddress = program.getProgram().getAddress();
        recipient.getTransfers().push(new Pair<>(sourceAddress, valueTransferred));
        var newContext = new ProgramContext(recipient);
        var inputData = program.getMemory().read(inOffset.intValue(), inSize.intValue());
        newContext.getCallData().clear();
        newContext.getCallData().write(0, inputData);
        process.getPrograms().push(newContext);
    }

    @Override
    public List<Integer> getStackAdds() {
        return List.of();
    }

    @Override
    public List<ProgramZone> getInstructionModifiers() {
        return List.of(STACK, PROGRAM_CONTEXT, MEMORY);
    }

    @Override
    public List<StackParameter> getStackParameters() {
        return List.of(RECIPIENT_ADDRESS, VALUE_TRANSFERRED, IN_OFFSET, IN_SIZE, OUT_OFFSET, OUT_SIZE);
    }

    @Override
    public String toString() {
        return "CALL";
    }
}

File: src/main/java/com/hileco/cortex/instructions/calls/CALL.kt
package com.hileco.cortex.instructions.calls

import com.hileco.cortex.constraints.expressions.Expression
import com.hileco.cortex.instructions.Instruction
import com.hileco.cortex.instructions.ProgramException
import com.hileco.cortex.instructions.ProgramException.Reason.CALL_RECIPIENT_MISSING
import com.hileco.cortex.instructions.ProgramException.Reason.STACK_UNDERFLOW
import com.hileco.cortex.instructions.StackParameter
import com.hileco.cortex.instructions.io.LOAD
import com.hileco.cortex.vm.ProgramZone
import com.hileco.cortex.vm.ProgramZone.*
import com.hileco.cortex.vm.concrete.ProgramContext
import com.hileco.cortex.vm.concrete.VirtualMachine
import com.hileco.cortex.vm.symbolic.SymbolicProgramContext
import com.hileco.cortex.vm.symbolic.SymbolicVirtualMachine
import java.math.BigInteger

class CALL : Instruction() {
    override val instructionModifiers: List<ProgramZone>
        get() = listOf(STACK, PROGRAM_CONTEXT, MEMORY)

    override val stackParameters: List<StackParameter>
        get() = listOf(RECIPIENT_ADDRESS, VALUE_TRANSFERRED, IN_OFFSET, IN_SIZE, OUT_OFFSET, OUT_SIZE)

    override fun execute(virtualMachine: VirtualMachine, programContext: ProgramContext) {
        if (programContext.stack.size() < 6) {
            throw ProgramException(STACK_UNDERFLOW)
        }
        val recipientAddress = BigInteger(programContext.stack.pop())
        val valueTransferred = BigInteger(programContext.stack.pop())
        val inOffset = BigInteger(programContext.stack.pop())
        val inSize = BigInteger(programContext.stack.pop())
        val outOffset = BigInteger(programContext.stack.pop())
        val outSize = BigInteger(programContext.stack.pop())
        programContext.returnDataOffset = outOffset
        programContext.returnDataSize = outSize
        val recipient = virtualMachine.atlas[recipientAddress] ?: throw ProgramException(CALL_RECIPIENT_MISSING)
        val sourceAddress = programContext.program.address
        recipient.transfers.push(sourceAddress to valueTransferred)
        val newContext = ProgramContext(recipient)
        val inputData = programContext.memory.read(inOffset.toInt() * LOAD.SIZE, inSize.toInt())
        newContext.callData.clear()
        newContext.callData.write(0, inputData)
        virtualMachine.programs.add(newContext)
    }

    override fun execute(virtualMachine: SymbolicVirtualMachine, programContext: SymbolicProgramContext) {
        if (programContext.stack.size() < 6) {
            throw ProgramException(STACK_UNDERFLOW)
        }
        val recipientAddress = programContext.stack.pop()
        val valueTransferred = programContext.stack.pop()
        val inOffset = programContext.stack.pop()
        val inSize = programContext.stack.pop()
        val outOffset = programContext.stack.pop()
        val outSize = programContext.stack.pop()
        if (inOffset != Expression.Value(0)
                || inSize != Expression.Value(0)
                || outOffset != Expression.Value(0)
                || outSize != Expression.Value(0)) {
            throw UnsupportedOperationException("Memory transfer is not supported for symbolic execution")
        }
        if (recipientAddress !is Expression.Value) {
            throw UnsupportedOperationException("Non-concrete address calling is not supported for symbolic execution")
        }
        val recipient = virtualMachine.atlas[recipientAddress.constant.toBigInteger()] ?: throw ProgramException(CALL_RECIPIENT_MISSING)
        val sourceAddress = programContext.program.address
        recipient.transfers.push(Expression.Value(sourceAddress.toLong()) to valueTransferred)
        val newContext = SymbolicProgramContext(recipient)
        virtualMachine.programs.add(newContext)
    }

    companion object {
        val RECIPIENT_ADDRESS = StackParameter("recipientAddress", 0)
        val VALUE_TRANSFERRED = StackParameter("valueTransferred", 1)
        val IN_OFFSET = StackParameter("inOffset", 2)
        val IN_SIZE = StackParameter("inSize", 3)
        val OUT_OFFSET = StackParameter("outOffset", 4)
        val OUT_SIZE = StackParameter("outSize", 5)
    }
}
